default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+            if stat[0] > self.enemy_weapon_strength:
+                self.enemy_weapon_strength = stat[0]
+
+    def convert_to_tuple(self):
+        """Create a tuple of this class' attributes (e.g. to use as key in dict).
+
+        :returns: tuple (minFuel,minSpeed,enemyDmg,enemyShield,enemyMineDmg)
+        """
+        tup = ("minFuel: %s" % self.minimum_fuel, "minSpeed: %s" % self.minimum_speed,
+               "enemyDmg: %s" % self.enemy_weapon_strength, "enemyShields: %s" % self.enemy_shields,
+               "enemyMineDmg: %s" % self.enemy_mine_dmg)
+        return tup
+
+
+class AIShipDesign(object):
+    """This class and its subclasses implement the building of a ship design and its rating.
+     Specialised Designs with their own rating system or optimizing algorithms should inherit from this class.
+
+    Member functions intended for external use:
+    optimize_design(): Returns the estimated optimum design according to the rating function
+    evaluate(): Returns a rating for the design as of the current state
+    update_hull(hullname): sets the hull used in the design
+    update_parts(partname_list): sets the parts used in the design
+    update_species(speciesname): sets the piloting species
+    update_stats(): calculates the stats of the design based on hull+parts+species
+    add_design(): Adds the shipdesign in the C++ part of the game
+
+    Functions which are to be overridden in inherited classes:
+    _rating_function()
+    _class_specific_filter()
+    _starting_guess()
+
+    For improved performance, maybe override _filling_algorithm() with a more specialised algorithm as well.
+    """
+    basename = "Default - Do not build"     # base design name
+    description = "Base class Ship type"    # design description
+    useful_part_classes = ALL_META_CLASSES  # only these parts are considered in the design process
+
+    filter_useful_parts = True              # removes any part not belonging to self.useful_part_classes
+    filter_inefficient_parts = False        # removes cost-inefficient parts (less capacity and less capacity/cost)
+
+    running_index = 1                       # running index used in the design name
+    running_index_needs_update = True       # To check if we need to update the running_index after a load.
+
+    def __init__(self):
+        """Make sure to call this constructor in each subclass."""
+        self.species = None         # name of the piloting species (string)
+        self.hull = None            # hull object (not hullname!)
+        self.partnames = []         # list of partnames (string)
+        self.parts = []             # list of actual part objects
+        self.attacks = {}           # {damage:count}
+        self.structure = 0
+        self.shields = 0
+        self.fuel = 0
+        self.speed = 0
+        self.stealth = 0
+        self.detection = 0
+        self.troops = 0
+        self.colonisation = -1      # -1 since 0 indicates an outpost (capacity = 0)
```

Comment:
If you take a look at the various derived subclasses, you will see that the remaining implementation is trivial and a pretty much no brainer by simply using the functionality the base class provides as blackbox.

Obviously, the algorithms used can potentially be more optimized for each design due to extended knowledge. This is generally speaking no problem as you can simply reimplement the methods but most likely for the most part not worth the trouble and complexification of the  code as the generic algorithm implemented in the most basic class should be capable of handling any not too complex ship design by defining an adequate _rating_function() and adjusting initial value (_starting_guess()) and maybe stating additional filters for performance improvements.